bndtools.core/src/bndtools/launch/BndContainerRuntimeClasspathEntryResolver.java

```diff
@@ -0,0 +1,98 @@
+package bndtools.launch;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.debug.core.ILaunchConfiguration;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.launching.IRuntimeClasspathEntry;
+import org.eclipse.jdt.launching.IRuntimeClasspathEntryResolver;
+import org.eclipse.jdt.launching.IVMInstall;
+import org.eclipse.jdt.launching.JavaRuntime;
+
+import aQute.bnd.build.Container;
+import aQute.bnd.build.Project;
+import aQute.lib.exceptions.Exceptions;
+import aQute.lib.strings.Strings;
+import bndtools.Activator;
+import bndtools.central.Central;
+
+public class BndContainerRuntimeClasspathEntryResolver implements IRuntimeClasspathEntryResolver {
+
+    @Override
+    public IRuntimeClasspathEntry[] resolveRuntimeClasspathEntry(IRuntimeClasspathEntry entry, ILaunchConfiguration configuration) throws CoreException {
+        IJavaProject project = entry.getJavaProject();
+
+        if (project == null) {
+            project = JavaRuntime.getJavaProject(configuration);
+        }
+
+        return resolveRuntimeClasspathEntry(entry, project);
+    }
+
+    @Override
+    public IVMInstall resolveVMInstall(IClasspathEntry entry) throws CoreException {
+        return null;
+    }
+
+    @Override
+    public IRuntimeClasspathEntry[] resolveRuntimeClasspathEntry(IRuntimeClasspathEntry entryToResolve, IJavaProject entryProject) throws CoreException {
+        if (entryToResolve == null || entryProject == null) {
+            return new IRuntimeClasspathEntry[0];
+        }
+
+        try {
+            return Central.bndCall(() -> {
+                List<IRuntimeClasspathEntry> resolved = new ArrayList<>();
+                Set<File> classpath = new HashSet<>();
+
+                Project project = Central.getProject(entryProject.getProject());
+
+                List<Container> containers = new ArrayList<>(project.getBuildpath());
+                containers.addAll(project.getTestpath());
```

Comment:
There seems to be a number of issues here. 

First, this work is already been done by the BndContainerInitializer and held in the BndContainer object:

```
IClasspathEntry[] entries = JavaModelManager.getJavaModelManager()
            .containerGet(javaProject, BndtoolsConstants.BND_CLASSPATH_ID).getClasspathEntries();
```

So we should not recompute it here since this code and BndContainerInitializer will diverge over time.

Second, the test path entries should only be added when it is a test launch, no? This code adds them all the time. We can use the new "test" attribute (`JavaCore.newClasspathAttribute("test", Boolean.TRUE.toString());` to identify entries which are testpath entries.

Reply:
These are IRuntimeClasspathEntry, not IClasspathEntry objects